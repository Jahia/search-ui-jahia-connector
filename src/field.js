const FieldType = {
    HIT: 'hit',
    NODE: 'node',
    REFERENCE_AS_VALUE: 'refValue',
    REFERENCE_AS_PATH: 'refPath'
};

class Field {
    /**
     *
     * @param type {FieldType} The type of field (ESHit or JCR)
     * @param name {string}  field name that should be returned in the response
     * @param alias {string} optional
     * @param useSnippet {boolean} configure if value is html based (snippet) or plain text (raw)
     */
    constructor(type, name, alias, useSnippet = false) {
        this.type = type;
        this.name = name;
        this.alias = alias;
        this.useSnippet = useSnippet;
    }

    resolveRequestField() {
        let fieldTemplate;
        switch (this.type) {
            case FieldType.HIT:
                fieldTemplate =  this.name;
                break;
            case FieldType.NODE:
                fieldTemplate = `${this.name.replace(':', '_')} : property(name: "${this.name}") {
                    value
                }`;
                break;
            case FieldType.REFERENCE_AS_PATH:
                fieldTemplate = `${this.name.replace(':', '_')} : property(name: "${this.name}") {
                    refNode {
                        path
                    }
                }`;
                break;
            case FieldType.REFERENCE_AS_VALUE:
                fieldTemplate = `${this.name.replace(':', '_')} : property(name: "${this.name}") {
                    refNode {
                        displayName
                    }
                }`;
                break;
        }
        return fieldTemplate;
    }

    resolveResponseField(hit, result) {
        let property = null;
        switch (this.type) {
            case FieldType.HIT:
                property = hit[this.name.replace(':', '_')];
                break;
            case FieldType.NODE:
                property = hit.node[this.name.replace(':', '_')].value;
                break;
            case FieldType.REFERENCE_AS_PATH:
                property = hit.node[this.name.replace(':', '_')].refNode.path;
                break;
            case FieldType.REFERENCE_AS_VALUE:
                property = hit.node[this.name.replace(':', '_')].refNode.displayName;
                break;
        }
        let field = {};
        if (this.useSnippet) {
            field.snippet = property;
        } else {
            field.raw = property;
        }
        result[this.alias ? this.alias : this.name.replace(':', '_')] = field;
    }
}

export {Field, FieldType};
