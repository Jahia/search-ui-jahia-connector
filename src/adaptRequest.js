import {Field, FieldType} from './field';
import {parse, print} from 'graphql';
import sort from './sort';
import facets from './facets';
import filters from './filters';

const buildFields = fields => {
    const fieldsConcatenated = {
        hitFields: '',
        nodeFields: ''
    };
    fields.forEach(field => {
        if (field.type === FieldType.HIT) {
            fieldsConcatenated.hitFields = `${fieldsConcatenated.hitFields},${field.resolveRequestField()}`;
        } else {
            fieldsConcatenated.nodeFields = `${fieldsConcatenated.nodeFields},${field.resolveRequestField()}`;
        }
    });
    return fieldsConcatenated;
};

function htmlEscape(str) {
    if (str) {
        return String(str)
            .replace(/&/g, '&amp;')
            .replace(/"/g, '&quot;')
            .replace(/'/g, '&#39;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;')
            .replace(/\\/g, '\\\\');
    }

    return '';
}

/**
 * Adapt the request from Search UI to Jahia Augmented Search
 * @param {RequestOptions} requestOptions the options for this request
 * @param {any} request the state of the current request
 * @param {any} queryConfig the query configuration as defined when initializing the App
 * @returns {string} the graphql query to be excuted on a Jahia backend
 */
export default function adaptRequest(requestOptions, request, queryConfig) {
    const graphQLOptions = {
        resultsPerPage: 5,
        current: 1,
        ...requestOptions,
        ...request
    };
    const resultFields = 'results' in queryConfig ? queryConfig.results.result_fields : queryConfig.result_fields;
    const resolvedRequestFields = buildFields(Object.keys(resultFields).reduce((acc, curr) => {
        const field = resultFields[curr];
        if (field instanceof Field) {
            acc.push(field);
        }

        return acc;
    }, []));

    return print(parse(`query {
        search(
            q: "${graphQLOptions.searchTerm !== undefined ? htmlEscape(graphQLOptions.searchTerm) : ''}",
            siteKeys: ["${graphQLOptions.siteKey}"],
            language: "${graphQLOptions.language}",
            workspace: ${graphQLOptions.workspace}
            ${filters(request, queryConfig, graphQLOptions)}
            ) {

            results(size: ${graphQLOptions.resultsPerPage},
                    page: ${graphQLOptions.current - 1}
                    ${sort(request)}
                    ) {
                totalHits
                took
                hits {
                    id
                    ${resolvedRequestFields.hitFields}
                    ${resolvedRequestFields.nodeFields}
                }
            }

            ${facets(request, queryConfig)}
        }
    }`));
}
