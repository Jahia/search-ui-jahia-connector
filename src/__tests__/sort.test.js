import adaptRequest from '../adaptRequest';
import {Field, FieldType} from '../field';

beforeEach(() => {
    global.queryConfig = {
        // eslint-disable-next-line camelcase
        result_fields: [
            new Field(FieldType.HIT, 'link'),
            new Field(FieldType.HIT, 'displayableName', 'title'),
            new Field(FieldType.HIT, 'excerpt', null, true),
            new Field(FieldType.HIT, 'score'),
            new Field(FieldType.NODE, 'jcr:created', 'created')
        ],
        facets: {}
    };
    global.state = {};
});

describe('Sort parameters tests', function () {
    const requestOptions = {
        siteKey: 'fake',
        language: 'fake',
        workspace: 'fake',
        nodeType: 'fake',
        functionScore: 'fake'
    };

    it('Query without sort', function () {
        const query = adaptRequest(requestOptions, state, queryConfig);
        expect(query).toMatchSnapshot();
    });

    it('Query with sort', function () {
        state.sortDirection = 'ASC';
        state.sortField = 'jcr:title';
        const query = adaptRequest(requestOptions, state, queryConfig);
        expect(query).toMatchSnapshot();
    });

    it('Query with incorrect sort field', function () {
        // Note that in this case sort parameters are ignored and the sort falls back to Relevance
        state.sortDirection = 'ASC';
        state.sortField = '';
        const query = adaptRequest(requestOptions, state, queryConfig);
        expect(query).toMatchSnapshot();
    });

    it('Query with incorrect sort direction', function () {
        // Note that in this case sort parameters are ignored and the sort falls back to Relevance
        state.sortDirection = '';
        state.sortField = 'jcr:title';
        const query = adaptRequest(requestOptions, state, queryConfig);
        expect(query).toMatchSnapshot();
    });
});
