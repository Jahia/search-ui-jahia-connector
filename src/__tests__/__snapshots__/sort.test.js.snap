// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort parameters tests Query with incorrect sort direction 1`] = `
"{
  search(
    q: \\"undefined\\"
    siteKeys: [\\"fake\\"]
    language: \\"fake\\"
    workspace: fake
    filters: {nodeType: {type: \\"fake\\"}}
  ) {
    results(size: 5, page: 0) {
      totalHits
      took
      hits {
        id
        link
        displayableName
        excerpt
        score
        created: property(name: \\"jcr:created\\")
      }
    }
  }
}
"
`;

exports[`Sort parameters tests Query with incorrect sort field 1`] = `
"{
  search(
    q: \\"undefined\\"
    siteKeys: [\\"fake\\"]
    language: \\"fake\\"
    workspace: fake
    filters: {nodeType: {type: \\"fake\\"}}
  ) {
    results(size: 5, page: 0) {
      totalHits
      took
      hits {
        id
        link
        displayableName
        excerpt
        score
        created: property(name: \\"jcr:created\\")
      }
    }
  }
}
"
`;

exports[`Sort parameters tests Query with sort 1`] = `
"{
  search(
    q: \\"undefined\\"
    siteKeys: [\\"fake\\"]
    language: \\"fake\\"
    workspace: fake
    filters: {nodeType: {type: \\"fake\\"}}
  ) {
    results(size: 5, page: 0, sortBy: {dir: ASC, field: \\"jcr:title\\"}) {
      totalHits
      took
      hits {
        id
        link
        displayableName
        excerpt
        score
        created: property(name: \\"jcr:created\\")
      }
    }
  }
}
"
`;

exports[`Sort parameters tests Query without sort 1`] = `
"{
  search(
    q: \\"undefined\\"
    siteKeys: [\\"fake\\"]
    language: \\"fake\\"
    workspace: fake
    filters: {nodeType: {type: \\"fake\\"}}
  ) {
    results(size: 5, page: 0) {
      totalHits
      took
      hits {
        id
        link
        displayableName
        excerpt
        score
        created: property(name: \\"jcr:created\\")
      }
    }
  }
}
"
`;
